### ISSUE TITLE
`sudo` auth does not work when requiring PIN

### ISSUE THREAD

**Issue by maximbaz** [Yubico/pam-u2f#144]
`/etc/pam.d/sudo` contains

```
auth		sufficient	pam_u2f.so	cue
```

Run command below:

```
$ pamu2fcfg -umaximbaz > ~/.config/Yubico/u2f_keys
```

Then `sudo -s` will work as expected, it will print "Please touch the device.", I touch it and I get sudo access.

Now do the same, but require PIN:

```
$ pamu2fcfg -umaximbaz -N > ~/.config/Yubico/u2f_keys
```

Running `sudo -s`:

1. First it asks "Please enter the PIN:", I enter it.
2. Then the message "Please touch the device." appears.
3. Instantly Yubikey stops flashing and I get the next message: `[sudo] password for maximbaz:`

In `journalctl` I see this:

```
sudo[1235117]: pam_unix(sudo:auth): conversation failed
kernel: audit: type=1100 audit(1583694800.396:4001): pid=1235117 uid=1000 auid=1000 ses=1 msg='op=PAM:authentication grantors=? acct="maximbaz" exe="/usr/bin/sudo" hostname=? addr=? >
audit[1235117]: USER_AUTH pid=1235117 uid=1000 auid=1000 ses=1 msg='op=PAM:authentication grantors=? acct="maximbaz" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/7 res=fai>
sudo[1235117]: pam_unix(sudo:auth): auth could not identify password for [maximbaz]
```

I'm on Arch Linux, latest master (https://github.com/Yubico/pam-u2f/commit/89412d932a24617ba6bb776988805dfbe663f795), YubiKey 5 Nano

Let me know if I can help with something, happy to help investigating.

**Comment by shibumi** [comment]
@maximbaz what is `cue` doing in your pam configuration?
And are you using a specific pamu2fcfg version in Arch Linux? If I do `pamu2fcfg --help` I don't see any `-N` flag.

**Comment by maximbaz** [comment]
`cue` is just something that makes the message "Please touch the device." appear when you run any `sudo` command, I've just tested, the issue is reproducible even without that parameter.

I'm using `AUR/pam_u2f-git`, waiting on #140 to update the community package 🙂

**Comment by shibumi** [comment]
Thanks!

**Comment by shibumi** [comment]
@maximbaz My guess id that you need to enable `pinverification=1` in your PAM. did you try this?

So something like this:
```
auth		sufficient	pam_u2f.so	cue pinverification=1
```

```
pamu2fcfg -umaximbaz -N > ~/.config/Yubico/u2f_keys
```

**Comment by maximbaz** [comment]
Interesting! Just tried, it doesn't help, it actually seems this `pinverification=1` is simply a way to enforce requesting PIN even if you didn't specify `-N` when running `pamu2fcfg`.

**Comment by shibumi** [comment]
Mh ok, sad. I thought it might work, then it's definitely an upstream bug.

**Comment by a-dma** [comment]
Thanks for the report, acknowledged. I'll look into it as soon as I have some time.

**Comment by SamSaffron** [comment]
Oh I was searching for the exact same thing 

I generated a cert in slot 9a, my guess is that there is that it want that public key somehow?

**Comment by a-dma** [comment]
Hi, sorry for the delay. This seems to be an issue with how the various options are set when asking for an assertion. It should be already fixed in the `ssh_format` branch. Would you be able to check it out and confirm?

**Comment by SamSaffron** [comment]
@a-dma it looks like `make check` is failing on the branch with a core dump

```
PASS: basic
debug(pam_u2f): util.c:941 (get_devices_from_authfile): The owner of the authentication file is neither  nor root
get_devices: get_devices.c:39: main: Assertion `rc == 1' failed.
/bin/sh: line 5: 35521 Aborted                 (core dumped) ${dir}$tst
FAIL: get_devices
=========================================================
1 of 2 tests failed
Please report to https://github.com/Yubico/pam-u2f/issues
=========================================================
make[2]: *** [Makefile:487: check-TESTS] Error 1
make[2]: Leaving directory '/home/sam/Source/pam-u2f/tests'
make[1]: *** [Makefile:614: check-am] Error 2
make[1]: Leaving directory '/home/sam/Source/pam-u2f/tests'
make: *** [Makefile:604: check-recursive] Error 1
```

**Comment by a-dma** [comment]
you need to chown the files in `tests/credentials` to root first.

**Comment by maximbaz** [comment]
I confirm that this bug is fixed in `ssh_format` branch 👍 

> you need to chown the files in tests/credentials to root first.

It would be nice to have the tests not require that, to be able to run them in jailed controlled environments while building a distro package for the next version. But not a requirement and definitely not a release blocker, if it can't be done, we can skip running tests as part of the build.

**Comment by SamSaffron** [comment]
I am still having trouble with the ssh format branch.

```
debug(pam_u2f): util.c:1088 (get_authenticators): Found key in authenticator 0
Please enter the PIN: 
debug(pam_u2f): pam-u2f.c:432 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:471 (pam_sm_authenticate): done. [Authentication failure]
```

It never passes the pin the yubikey cause I can just randomly try strings and the pin attempt count does not go down.

Confirm I have `e2a1e58` installed.

Any idea on how I debug further.

Confirm chowning tests/credentials makes the tests pass

**Comment by a-dma** [comment]
@maximbaz I agree, but those tests only check that the credential file can be correctly parsed. I wanted to test the function with as little modification as possible, which means there's a bunch of hops you need to jump through to open those files. This solution seems to work fine on Travis. Generally speaking, automating tests for this project is a bit of a nightmare,

@SamSaffron weird, it seems to work fine for me, if I pass the wrong PIN I fail the authentication and lose one attempt in the retry counter. I saw you mentioning slot 9a in an earlier message, what are you using to set the PIN and looking at the retry counter?

Thank you both for helping with testing.

**Comment by SamSaffron** [comment]
Hi @a-dma  yeah no luck, I even tried a second Yubikey

(replaced stuff that looks like keys with XYZ) 

```
debug(pam_u2f): pam-u2f.c:110 (parse_cfg): called.
debug(pam_u2f): pam-u2f.c:111 (parse_cfg): flags 32768 argc 1
debug(pam_u2f): pam-u2f.c:113 (parse_cfg): argv[0]=debug
debug(pam_u2f): pam-u2f.c:115 (parse_cfg): max_devices=0
debug(pam_u2f): pam-u2f.c:116 (parse_cfg): debug=1
debug(pam_u2f): pam-u2f.c:117 (parse_cfg): interactive=0
debug(pam_u2f): pam-u2f.c:118 (parse_cfg): cue=0
debug(pam_u2f): pam-u2f.c:119 (parse_cfg): nodetect=0
debug(pam_u2f): pam-u2f.c:120 (parse_cfg): userpresence=-1
debug(pam_u2f): pam-u2f.c:121 (parse_cfg): userverification=-1
debug(pam_u2f): pam-u2f.c:122 (parse_cfg): pinverification=-1
debug(pam_u2f): pam-u2f.c:123 (parse_cfg): manual=0
debug(pam_u2f): pam-u2f.c:124 (parse_cfg): nouserok=0
debug(pam_u2f): pam-u2f.c:125 (parse_cfg): openasuser=0
debug(pam_u2f): pam-u2f.c:126 (parse_cfg): alwaysok=0
debug(pam_u2f): pam-u2f.c:127 (parse_cfg): sshformat=0
debug(pam_u2f): pam-u2f.c:128 (parse_cfg): authfile=(null)
debug(pam_u2f): pam-u2f.c:130 (parse_cfg): authpending_file=(null)
debug(pam_u2f): pam-u2f.c:132 (parse_cfg): origin=(null)
debug(pam_u2f): pam-u2f.c:133 (parse_cfg): appid=(null)
debug(pam_u2f): pam-u2f.c:134 (parse_cfg): prompt=(null)
debug(pam_u2f): pam-u2f.c:192 (pam_sm_authenticate): Origin not specified, using "pam://slaptop"
debug(pam_u2f): pam-u2f.c:203 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://slaptop)
debug(pam_u2f): pam-u2f.c:215 (pam_sm_authenticate): Maximum devices number not set. Using default (24)
debug(pam_u2f): pam-u2f.c:233 (pam_sm_authenticate): Requesting authentication for user sam
debug(pam_u2f): pam-u2f.c:244 (pam_sm_authenticate): Found user sam
debug(pam_u2f): pam-u2f.c:245 (pam_sm_authenticate): Home directory for sam is /home/sam
debug(pam_u2f): pam-u2f.c:259 (pam_sm_authenticate): Variable XDG_CONFIG_HOME is not set. Using default value ($HOME/.config/)
debug(pam_u2f): pam-u2f.c:296 (pam_sm_authenticate): Using authentication file /home/sam/.config/Yubico/u2f_keys
debug(pam_u2f): pam-u2f.c:331 (pam_sm_authenticate): Dropping privileges
debug(pam_u2f): pam-u2f.c:337 (pam_sm_authenticate): Switched to uid 1000
debug(pam_u2f): util.c:160 (parse_native_format): Authorization line: sam:XYZ,es256,+presence+pin
debug(pam_u2f): util.c:165 (parse_native_format): Matched user: sam
debug(pam_u2f): util.c:201 (parse_native_format): KeyHandle for device number 1: XYZ
debug(pam_u2f): util.c:225 (parse_native_format): publicKey for device number 1: XYZ
debug(pam_u2f): util.c:250 (parse_native_format): COSE type for device number 1: es256
debug(pam_u2f): util.c:272 (parse_native_format): Attributes for device number 1: +presence+pin
debug(pam_u2f): util.c:979 (get_devices_from_authfile): Found 1 device(s) for user sam
debug(pam_u2f): pam-u2f.c:347 (pam_sm_authenticate): Restored privileges
debug(pam_u2f): pam-u2f.c:396 (pam_sm_authenticate): Using file '/var/run/user/1000/pam-u2f-authpending' for emitting touch request notifications
get_key_val: key_len=4
get_key_val: key_len=4
debug(pam_u2f): util.c:1152 (do_authentication): Device max index is 1
debug(pam_u2f): util.c:1184 (do_authentication): Attempting authentication with device number 1
debug(pam_u2f): util.c:1206 (do_authentication): Key handle: XYZ
debug(pam_u2f): util.c:1311 (do_authentication): Challenge: wKUNnjpRExNtFNW5TJoBrbyxRflJo/lKiJsYEHmb+N0=
debug(pam_u2f): util.c:1047 (get_authenticators): Working with 1 authenticator(s)
debug(pam_u2f): util.c:1051 (get_authenticators): Checking whether key exists in authenticator 0
debug(pam_u2f): util.c:1061 (get_authenticators): Authenticator path: /dev/hidraw3
fido_tx: d=0x55c738f1bfa0, cmd=0x86, buf=0x55c738f1bfa0, count=8
  16 72 c7 50 27 33 83 a6 
fido_rx: initiation frame at 0x7ffc23e17730, len 64
  ff ff ff ff 86 .... 
fido_tx: d=0x55c738f1bfa0, cmd=0x83, buf=0x55c738eeaac2, count=138
  00 02 07 00 00 00 81 ff ff ff ff ff ff ff ff ff 
  ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
  ff ff ff ff ff ff ff 86 .... 
fido_rx: initiation frame at 0x7ffc23e173f0, len 64
  0d 00 53 01 83 ....
u2f_authenticate_single: checking for key existence only
u2f_authenticate: u2f_authenticate_single
debug(pam_u2f): util.c:1088 (get_authenticators): Found key in authenticator 0
Please enter the PIN:

get_key_val: key_len=4
get_key_val: key_len=4
debug(pam_u2f): pam-u2f.c:432 (pam_sm_authenticate): do_authentication returned -2
debug(pam_u2f): pam-u2f.c:471 (pam_sm_authenticate): done. [Authentication failure]

```

My pin clearly works cause:

```
sam@slaptop Yubico % yubico-piv-tool -averify-pin -Pgoodpin
Successfully verified PIN.
sam@slaptop Yubico % yubico-piv-tool -averify-pin -Pbadpin
Pin verification failed, 2 tries left before pin is blocked.
```

**Comment by a-dma** [comment]
You're using the wrong program, `yubico-piv-tool` exercises the PIV application on a YubiKey, this project uses the U2F/FIDO2 application, those are completely separate and unrelated applications. So are their PINs.

You can use `ykman fido set-pin` to set a PIN on the FIDO2 application and `ykman fido info` to see how many attempts you have left (`ykman` is from the YubiKey Manager project).

**Comment by SamSaffron** [comment]
@a-dma  :cry:  :crying_cat_face:  turns out I have a YubiKey NEO (actually 2 of them) and neither of them are fido 2, hence my confusion here. 

I guess we should improve the readme a bit to clarify that you need a YubiKey 5 for this feature.

**Comment by a-dma** [comment]
Ah, I see.
Yes, you can still use `pam-u2f`, but only FIDO2 supports PIN. That should maybe be handled more gracefully.

**Comment by a-dma** [comment]
Those changes are now in master and have been released.
